### PR TITLE
Adds Service Versioning vocabulary

### PR DESCRIPTION
**Pull request by christopherabrooks**
Initial files for ServiceVersioning vocabulary

### CODE REVIEW

File: OData.Community.ServiceVersioning.V1.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+
+```http
+GET /service/Customers?api-version=7.2%2Cisvsolution1%2F5.0%2Cisvsolution2%2F3.1 HTTP/1.1
+Host: odata.org
+OData-Version: 4.0 
+OData-MaxVersion: 4.0 
+Content-Type: application/json
+```
+
+###3.3	Service Responses
```
Comment (ghost) [inline OData.Community.ServiceVersioning.V1.md:443]: > +When services communicate that their Version Information is required on client calls, but a client calls without it, the service is expected to respond with HTTP error status such as `400 Bad Request` with a message indicating that the version header or query string parameter is required.

To me, this wording suggests that a client _MUST_ understand the semantics of this vocabulary in order to interact with a service that uses the vocabulary. This seems to contradict the intent of the spec, which suggests that header and vocabulary extensibility should be optional. Relevant text from the spec below: (Emphasis mine) 

> 6.5 Header Field Extensibility
> OData defines semantics around certain HTTP request and response headers. Services that support a version of OData conform to the processing requirements for the headers defined by this specification for that version.
> Individual services may define custom headers. These headers MUST NOT begin with OData . **Custom headers SHOULD be optional when making requests to the service.** A service MUST NOT require the client to understand custom headers to accurately interpret the response.

http://docs.oasis-open.org/odata/odata/v4.0/os/part1-protocol/odata-v4.0-os-part1-protocol.html#_Toc372793605 

An acceptable behavior when a client does not present version information would just be to expose the latest version of the service. That seems better to me than throwing a 400 and it's more in line with the spec.

File: OData.Community.ServiceVersioning.V1.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          <Annotation Term="Core.Description" String="Indicates whether the service requires that the version be included in client requests."/>
+        </Property>
+        <Property Type="Edm.String" Name="VersionHeaderName" Nullable="true">
+          <Annotation Term="Core.Description" String="The name of the header clients should use to communicate version information."/>
+        </Property>
+        <Property Type="Edm.String" Name="VersionQueryStringParameterName" Nullable="true">
+          <Annotation Term="Core.Description" String="The name of the query string parameter clients should use to communicate version information."/>
+        </Property>
+      </ComplexType>
+
```
Comment (ghost) [inline OData.Community.ServiceVersioning.V1.xml]: nit: This spacing probably isn't nescessary

File: OData.Community.ServiceVersioning.V1.md (hunk 1 above)
Comment (christopherabrooks) [inline OData.Community.ServiceVersioning.V1.md:443]: This behavior would depend on the "Required" property.  The default is not required, and the expectations would be as you describe.  

But for some services, having version resilient client calls is a requirement, and so the point of this vocabulary is to allow those services to communicate this requirement in a standard way.

File: OData.Community.ServiceVersioning.V1.xml (hunk 2 above)
Comment (christopherabrooks) [inline OData.Community.ServiceVersioning.V1.xml]: fixed - thanks for the feedback.

File: OData.Community.ServiceVersioning.V1.md (hunk 1 above)
Comment (ralfhandl) [inline OData.Community.ServiceVersioning.V1.md:443]: Strictly speaking a service that requires non-standard headers isn't a compliant OData service, so having a "standard way" of making a service non-standard is a contradiction in itself.

I strongly recommend removing the "Required" property.

Comment (christopherabrooks) [inline OData.Community.ServiceVersioning.V1.md:443]: I am open to a note saying that using "required" with only supporting version in a header would represent a non-compliant service (for the reason that you mention).

The more typical pattern would be for a service to "Require" that a version be passed with each call through a query string parameter, but potentially (alternatively) also accepting a header.

Not allowing a service to express that passing a version is required by that service (and will always result in an error result from the service) defeats the purpose of this annotation.

Comment (ralfhandl) [inline OData.Community.ServiceVersioning.V1.md:443]: Not allowing a client to interact with an OData service in a a standard-compliant way seems to defeat the purpose of OData. 

Just make your service work without the proprietary header or query option by either exposing the latest model version or the oldest supported model version. Clients that need "version safety" are special and can very well send special requests. 

This solves your problem without denying standard clients access to your service.

**Pull request closed without merging**